src/architecture/merkledb.md

```diff
@@ -0,0 +1,348 @@
+
+The pool is stored in a separate column family. Key in this column family is
+the object address presented in bytes, value is the
+[`IndexMetadata`][index-metadata] structure, which stores the object
+identifier and some object metadata.
+
+The object metadata can store various information about the internal state of
+the object. For example, for [`ListIndex`], the metadata stores the length of
+the list.
+
+An object key in RocksDB consists of two parts:
+
+- an **object identifier** obtained from the object pool
+- an **index of an item** inside the object.
+
+Both parts of the key are encoded as big-endian.
+
+!!! note "Example"
+    Suppose we have a list with the address `(" exchange.crypto "," BTC ")` in
+    which we put one value, for example `7865`. The pool assigns identifier
+    `3` to this list when we create it. In the database the identifier looks
+    like this - `0x0000000000000003`.
+
+    Since `7865` is a single value in the created list, its index in the
+    database looks like this - `0x0000000000000000`.
+
+    Thus, the whole key in the database will look as follows (in HEX) -
+    `0x00000000000000030000000000000000 : 0x302E3030303139`.
+
+### Key Sorting and Iterators
+
+MerkleDB objects support iterating over object contents:
+
+- items for lists and sets
+- keys, values or key-value pairs in case of map objects.
+
+Such iterators use key ordering of the low-level key-value storage to
+determine the iteration order. Namely, keys are lexicographically ordered
+according to their binary serializations.
+
+## View Layer
+
+Exonum introduces additional layer over the database to handle transaction
+and block atomicity.
+
+### Patches
+
+[Patch][patch] is a set of serial changes that should be applied to
+the low-level storage atomically. A patch may include two types of
+operations: put a value addressed by a key, or delete a value by a key.
+
+### Snapshots
+
+[Snapshot][snapshot] fixes the storage state at the moment of snapshot
+creation and provides a read-only API to it. Even if the storage state
+is updated, the snapshot still refers to the old content of the stored objects.
+
+### Forks
+
+[Forks][fork] implement the same interfaces as the database underneath,
+transparently wrapping the real data storage state, and some
+additional changes. Every fork is based on the storage snapshot. From
+the outer point of view, the changes are eagerly applied to the data
+storage; however, these changes are stored directly in the fork and may
+be easily rolled back. You can create multiple mutable objects from one
+immutable reference to the fork. Moreover, there may be different forks of
+the same database snapshot.
+
+Forks are used during transaction and block processing. A fork
+[is successively passed](transactions.md#execute)
+to each transaction in the block to accumulate changes produced by the
+transactions, in a [patch](#patches). If one of the transactions in the block
```

Comment:
Nit: I'd say that the changes are accumulated in a fork (not a patch), i.e., remove "...in a patch." A patch is a result of _finalizing_ a fork.

Reply:
done